File: webproxy/webproxy.py
#!/usr/bin/env python3

# File: webproxy.py
# Description: cn-adta6 socket programming assignment 4, python version
#              a multi-threaded web proxy
# Author: Burgess Wong
# Created Date: 2013-11-11

import io
import os
import socket

def get_start_fields(start_line):
    return tuple(start_line.decode('ASCII').split(' '))

def get_header_fields(header_lines):
    header_fields = {}
    rest_lines = header_lines.decode('ASCII')
    while len(rest_lines) > 0:
        current_line, sep, rest_lines = rest_lines.partition('\r\n')
        name, sep, value = current_line.partition(': ')
        header_fields[name] = value

    return header_fields

def get_hostname_from_url(url):
    protocol, sep, rest_url = url.partition('://')
    hostname, sep, rest_url = rest_url.partition('/')
    return hostname
    
def separate_http_message(message):
    start_line, sep, rest_message = message.partition(b'\r\n')
    header_lines, sep, body = rest_message.partition(b'\r\n\r\n')
    header_lines += b'\r\n'
    
    return start_line, header_lines, body

def do_proxy(client_sock, pid):
    # get the request from the client
    request = client_sock.recv(4096)

    # get request line, header lines and entity body
    request_line, header_lines, body = separate_http_message(request)

    # get header fields dictionary
    header_fields = get_header_fields(header_lines)
    
    # remove Connection and Proxy-Connection from header fields
    if 'Connection' in header_fields:
        del header_fields['Connection']
    if 'Proxy-Connection' in header_fields:
        del header_fields['Proxy-Connection']

    # connect to the server
    server_name = header_fields['Host'] if 'Host' in header_fields else \
                  get_hostname_from_url(get_start_fields(request_line)[1])
    server_port = 80
    print('---Connect to the server %s, pid=%d' % (server_name, pid))
    server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_sock.connect((server_name, server_port))

    # send the new request, assuming that the whole request is less than 4096 bytes
    print('---Send the new request to the server, pid=%d' % pid)
    new_header_lines = ''
    for header_name in iter(header_fields):
        new_header_lines += '%s: %s\r\n' % (header_name, header_fields[header_name])
    server_sock.send(request_line + b'\r\n' + new_header_lines.encode('ASCII') + b'\r\n' + body)
    
    # get and send back the response
    print('---Send the response back to the client, pid=%d' % pid)
    response = server_sock.recv(4096)
    while len(response) > 0:
        client_sock.send(response)
        response = server_sock.recv(4096)

    # close server socket
    print('---Close server socket, pid=%d' % pid)
    server_sock.close()


        
if __name__ == '__main__':

    # initialize the socket as a server
    proxy_ip = ''
    proxy_port = 8080
    proxy_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    proxy_sock.bind((proxy_ip, proxy_port))
    proxy_sock.listen(5)

    parent_pid = os.getpid()
    
    # infinite loop waiting for and processing requests
    while 1:
        client_sock, client_address = proxy_sock.accept()

        print('---Request comes, IP=' + str(client_address[0]))
        
        # process the request in the subprocess
        if os.fork() == 0:
            pid = os.getpid()
            print('---Process the request in a subprocess, pid=%d' % pid)

            print('---Close proxy socket in the subprocess, pid=%d' % pid)
            proxy_sock.close()

            do_proxy(client_sock, pid)

            print('---Close the client socket in the subprocess, pid=%d' % pid)
            client_sock.close()

            print('---Subprocess exits, pid=%d' % pid)
            exit()

        print('---Close the client socket in the parent process')
        client_sock.close()

        

File: webproxy/webproxy.c
/**
 * File: webproxy.c
 * Description: cn-adta6 socket programming assignment 4, python version
 *              a multi-threaded web proxy
 * Author: Burgess Wong
 * Created Date: 2013-11-12
 */

#include "general_operation.h"

#define HTTP_PORT 80
#define PROXY_PORT 8080

struct header_field {
    const char *name, *value;
    int nlen, vlen;
    struct header_field *next;
};

inline int min(int i, int j) {
    return i < j ? i : j;
}

inline int max(int i, int j){
    return i > j ? i : j;
}

void get_start_fields(const char *start_line, int sllen, const char **first_field, int *flen, \
		      const char **second_field, int *slen, const char **third_field, int *tlen) {
    *first_field = start_line;
    *second_field = strstr(*first_field, " ");
    *second_field += 1;
    *flen = *second_field - *first_field - 1;
    *third_field = strstr(*second_field, " ");
    *third_field += 1;
    *slen = *third_field - *second_field - 1;
    *tlen = sllen - *flen - *slen - 2;
}

struct header_field *get_header_fields(const char *header_lines, int hlen) {
    int len;
    const char *iter_line, *rest_lines;
    struct header_field header_fields, *header_fields_tail, *new_field;

    len = 0;
    header_fields.next = NULL;
    header_fields_tail = &header_fields;
    rest_lines = header_lines;

    while (1) {
	iter_line = rest_lines;
	rest_lines = strstr(rest_lines, "\r\n") + 2;

	new_field = calloc(1, sizeof(struct header_field));
	new_field->name = iter_line;
	new_field->value = strstr(iter_line, ": ") + 2;
	new_field->nlen = new_field->value - new_field->name - 2;
	new_field->vlen = rest_lines - new_field->value - 2;
	new_field->next = NULL;

	header_fields_tail->next = new_field;
	header_fields_tail = new_field;

	len += rest_lines - iter_line;

	if (len == hlen)
	    break;
    }

    return header_fields.next;
}

void get_host_from_url(const char *url, const char **host, int *hostlen) {
    const char *rest;
    *host = strstr(url, "://");
    *host += 3;
    rest = strstr(*host, "/");
    *hostlen = rest - *host;
}

const struct header_field *get_header_field(const struct header_field *header_fields, const char *name) {
    const struct header_field *iter_field;

    for (iter_field = header_fields; iter_field; iter_field = iter_field->next)
	if (strncmp(name, iter_field->name, iter_field->nlen) == 0)
	    return iter_field;

    return NULL;
}

void remove_header_field(struct header_field **p_header_fields, const char *name) {
    struct header_field **p_iter_field, *current_field;
    p_iter_field = p_header_fields;
    while (*p_iter_field) {
	current_field = *p_iter_field;
	if (strncmp(name, current_field->name, current_field->nlen) == 0) {
	    *p_iter_field = current_field->next;
	    free(current_field);
	    return;
	} else {
	    p_iter_field = &current_field->next;
	}
    }
}

void separate_http_message(const char *message, int mlen, const char **start_line, int *slen, \
                           const char **header_lines, int *hlen, const char **body, int *blen) {
    *start_line = message;
    *header_lines = strstr(*start_line, "\r\n");
    *header_lines += 2;
    *slen = *header_lines - *start_line - 2;
    *body = strstr(*header_lines, "\r\n\r\n");
    *body += 4;
    *hlen = *body - *header_lines - 2;
    *blen = mlen - *slen - *hlen - 4;
}

char *write_to_buf(char *buf, const char *content, int clen) {
    memcpy(buf, content, clen);

    return buf + clen;
}

void do_proxy(int clifd, const struct sockaddr_in *cliaddr, int pid) {
    /* get the request from the client */
    char *request;
    int request_len;
    request = calloc(MAXLINE, sizeof(char));
    request_len = Read(clifd, request, MAXLINE);

    /* get the request line, header lines and entity body */
    const char *request_line, *header_lines, *body;
    int rlen, hlen, blen;
    separate_http_message(request, request_len, &request_line, &rlen, &header_lines, &hlen, &body, &blen);
    
    /* get header fields */
    struct header_field *header_fields;
    header_fields = get_header_fields(header_lines, hlen);

    /* remove Connection and Proxy-Connection from header fields */
    remove_header_field(&header_fields, "Connection");
    remove_header_field(&header_fields, "Proxy-Connection");

    /* get the hostname of the server */
    const struct header_field *host_header_field;
    char *host;
    const char *p_host, *method, *url, *version;
    int hostlen, methodlen, urllen, versionlen;
    if ((host_header_field = get_header_field(header_fields, "Host")) == NULL) {
	get_start_fields(request_line, rlen, &method, &methodlen, &url, &urllen, &version, &versionlen);
	get_host_from_url(url, &p_host, &hostlen);
    } else {
	p_host = host_header_field->value;
	hostlen = host_header_field->vlen;
    }
    host = calloc(1, hostlen + 1);
    memcpy(host, p_host, hostlen);

    /* connect to the server */
    int servfd;
    struct sockaddr_in servaddr;
    socklen_t servaddrlen;
    struct addrinfo hints, *result;
    fprintf(stdout, "---pid=%d, Connect to the server\n", pid);
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;
    getaddrinfo(host, "http", &hints, &result);
    memcpy(&servaddr, result->ai_addr, sizeof(servaddr));
    servaddrlen = result->ai_addrlen;
    freeaddrinfo(result);

    servfd = Socket(AF_INET, SOCK_STREAM, 0);
    Connect(servfd, (struct sockaddr *)&servaddr, servaddrlen);

    /* assemble the new request, assuming that the whole request is less than 4096 bytes */
    char *new_request, *bufpos;
    int new_request_len;
    struct header_field *iter_header_field;
    new_request = calloc(MAXLINE, sizeof(char));
    new_request_len = 0;
    bufpos = new_request;
    bufpos = write_to_buf(bufpos, request_line, rlen);
    bufpos = write_to_buf(bufpos, "\r\n", 2);
    new_request_len += rlen + 2;
    for (iter_header_field = header_fields; iter_header_field; iter_header_field = iter_header_field->next) {
	bufpos = write_to_buf(bufpos, iter_header_field->name, iter_header_field->nlen);
	bufpos = write_to_buf(bufpos, ": ", 2);
	bufpos = write_to_buf(bufpos, iter_header_field->value, iter_header_field->vlen);
	bufpos = write_to_buf(bufpos, "\r\n", 2);
	new_request_len += iter_header_field->nlen + iter_header_field->vlen + 4;
    }
    bufpos = write_to_buf(bufpos, "\r\n", 2);
    bufpos = write_to_buf(bufpos, body, blen);
    new_request_len += blen + 2;

    /* send the new request to the server */
    fprintf(stdout, "---pid=%d, Send the new request to the server\n", pid);
    Write(servfd, new_request, new_request_len);

    /* get and send back the response */
    char *response;
    int response_len;
    fprintf(stdout, "---pid=%d, Send the response back to the client\n", pid);
    response = calloc(MAXLINE, sizeof(char));

    while ((response_len = Read(servfd, response, MAXLINE)) > 0) {
	fprintf(stdout, "---pid=%d, %d bytes received from the server\n", pid, response_len);
	Write(clifd, response, response_len);
    }
    
    /* close the server socket */
    fprintf(stdout, "---pid=%d, Close the server socket\n", pid);
    close(servfd);

    /* release memory */
    struct header_field *next_header_field;
    free(request);
    free(new_request);
    free(response);
    free(host);
    for (iter_header_field = header_fields; iter_header_field; iter_header_field = next_header_field) {
	next_header_field = iter_header_field->next;
	free(iter_header_field);
    }
}

int main(int argc, char *argv[]) {
    int proxyfd, clifd;
    struct sockaddr_in proxyaddr, cliaddr;
    socklen_t cliaddrlen;

    /* initialize the socket as a server */
    proxyfd = Socket(AF_INET, SOCK_STREAM, 0);
    bzero(&proxyaddr, sizeof(proxyaddr));
    proxyaddr.sin_family = AF_INET;
    proxyaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    proxyaddr.sin_port = htons(PROXY_PORT);
    Bind(proxyfd, (struct sockaddr *)&proxyaddr, sizeof(proxyaddr));
    Listen(proxyfd, 5);

    /* infinite loop waiting for and processing requests */
    char cliaddrip[INET_ADDRSTRLEN];
    int pid;
    while (1) {
	cliaddrlen = sizeof(cliaddr);
	clifd = Accept(proxyfd, (struct sockaddr *)&cliaddr, &cliaddrlen);
	Inet_ntop(AF_INET, &cliaddr.sin_addr, cliaddrip, INET_ADDRSTRLEN);

	fprintf(stdout, "---Request comes, IP=%s\n", cliaddrip);

	/* process the request in a subprocess */
	if ((pid=fork()) == 0) {
	    pid = getpid();
	    fprintf(stdout, "---pid=%d, Process the request in a subprocess\n", pid);

	    fprintf(stdout, "---pid=%d, Close proxy socket in the subprocess\n", pid);
	    Close(proxyfd);

	    do_proxy(clifd, &cliaddr, pid);

	    fprintf(stdout, "---pid=%d, Close the client socket in the subprocess\n", pid);
	    Close(clifd);

	    fprintf(stdout, "---pid=%d, Subprocess exits\n", pid);
	    exit(0);
	}

	fprintf(stdout, "---Close the client socket in the parent process\n");
	Close(clifd);
    }

    return 0;
}
    
